x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/CompositeAggregationDataExtractor.java

```diff
@@ -0,0 +1,194 @@
+        SearchResponse searchResponse = executeSearchRequest(searchRequest);
+        LOGGER.trace(() -> new ParameterizedMessage("[{}] Search composite response was obtained", context.jobId));
+        timingStatsReporter.reportSearchDuration(searchResponse.getTook());
+        Aggregations aggregations = searchResponse.getAggregations();
+        if (aggregations == null) {
+            return null;
+        }
+        CompositeAggregation compositeAgg = aggregations.get(compositeAggregationBuilder.getName());
+        if (compositeAgg == null || compositeAgg.getBuckets().isEmpty()) {
+            return null;
+        }
+        afterKey = compositeAgg.afterKey();
+
+        return aggregations;
+    }
+
+    protected SearchResponse executeSearchRequest(ActionRequestBuilder<SearchRequest, SearchResponse> searchRequestBuilder) {
+        return ClientHelper.executeWithHeaders(context.headers, ClientHelper.ML_ORIGIN, client, searchRequestBuilder::get);
+    }
+
+    private InputStream processAggs(Aggregations aggs) throws IOException {
+        AggregationToJsonProcessor aggregationToJsonProcessor = new AggregationToJsonProcessor(
+            context.timeField,
+            context.fields,
+            context.includeDocCount,
+            context.start,
+            context.compositeAggDateHistogramGroupSourceName
+        );
+        LOGGER.trace(() -> new ParameterizedMessage(
+            "[{}] got [{}] composite buckets",
+            context.jobId,
+            ((CompositeAggregation)aggs.get(compositeAggregationBuilder.getName())).getBuckets().size()
+        ));
+        aggregationToJsonProcessor.process(aggs);
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        final boolean hasAfterKey = afterKey != null && (afterKey.get(context.compositeAggDateHistogramGroupSourceName) instanceof Long);
+        final Long nextBucket = hasAfterKey ? (Long)afterKey.get(context.compositeAggDateHistogramGroupSourceName) + interval : null;
+        boolean cancellable = aggregationToJsonProcessor.writeAllDocsCancellable(
+            timestamp -> {
+                if (isCancelled) {
+                    // If we have not processed a single composite agg page yet and we are cancelled
+                    // We should not process anything
+                    if (hasAfterKey == false) {
+                        return true;
+                    }
+                    // If the current timestamp is part of the next bucket given our current afterKey, the process can be cancelled
+                    // as it has passed into the next date_histogram bucket in the composite aggregation paging
+                    return timestamp >= nextBucket;
+                }
+                return false;
+            }, outputStream);
+        // If the process is canceled and cancelable, then we can indicate that there are no more buckets to process.
+        if (isCancelled && cancellable) {
+            hasNext = false;
+        }
```

Comment:
There is a bug here that I am fixing. This logic could cause the job to never stop depending on how it sees pages + docs.

Really, what we should do is on the first timestamp we see when cancelled, round up to the next bucket. Then keep that value in memory and allow cancellation once we pass that value. Working on the fix + figuring out a way to test.

Reply:
The latest commit fixes this and adds some nice functional tests.